pkg/operator/controller/instance_controller.go

```diff
@@ -0,0 +1,742 @@
+									MountPath: dpapp.CertificateFile,
+									SubPath:   "cert",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "tls",
+									MountPath: dpapp.KeyFile,
+									SubPath:   "key",
+									ReadOnly:  true,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return r.createOrUpdateResource(ctx, dpDeployment)
+}
+
+// setService sets up a k8s service.
+func (r *InstanceReconciler) setService(ctx context.Context, name string, port uint16) error {
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: r.clNamespace},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Protocol: corev1.ProtocolTCP,
+					Port:     int32(port),
+				},
+			},
+			Type:     corev1.ServiceTypeClusterIP,
+			Selector: map[string]string{"app": name},
+		},
+	}
+
+	return r.createResource(ctx, service)
+}
+
+// setService sets up k8s a persistent volume claim for the.
+func (r *InstanceReconciler) setPVC(ctx context.Context, name string) error {
+	// Create the PVC for cl-controlplane
+	controlplanePVC := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: r.clNamespace,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+			Resources: corev1.VolumeResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: resource.MustParse("100Mi"),
+				},
+			},
+		},
+	}
+
+	return r.createResource(ctx, controlplanePVC)
+}
+
+// setRules sets up k8s ClusterRule and ClusterRoleBinding for the controlplane.
+func (r *InstanceReconciler) setRules(ctx context.Context, name string) error {
+	// Create the ClusterRole for the controlplane.
+	clusterRole := &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: r.clNamespace,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"services"},
+				Verbs:     []string{"create", "delete", "update"},
+			},
+			{
+				APIGroups: []string{""},
+				Resources: []string{"pods"},
+				Verbs:     []string{"get", "list", "watch"},
+			},
+		},
+	}
+
+	err := r.createResource(ctx, clusterRole)
+	if err != nil {
+		return err
+	}
+
+	// Create ClusterRoleBinding for the controlplane.
+	clusterRoleBinding := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: r.clNamespace,
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "ClusterRole",
+			Name:     name,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      "default",
+				Namespace: "default",
+			},
+		},
+	}
+	return r.createResource(ctx, clusterRoleBinding)
+}
+
+// setNameSpace sets up k8s ClusterRule and ClusterRoleBinding for the controlplane.
+func (r *InstanceReconciler) setNameSpace(ctx context.Context) error {
```

Comment:
- Comment seems not aligned with the actual code.
- I think you can assume the destination namespace always exists and fail if it does not (mark the conditions on the Instance object to failed or invalid as well). I would consider adding the destination namespace to the Spec (default value is `clusterlink-system` when not set. It will make life easier and allow some customization to users without much work on our part.